### PR TITLE
Visible source control providers API, #53892

### PR DESCRIPTION
**Pull request by RMacfarlane**
Adds the concept of `visibleSourceControls` to the API, https://github.com/Microsoft/vscode/issues/53892

### CODE REVIEW

**Review by joaomoreno — APPROVED** [review]
Looks pretty good to me, though I haven't run it.

File: src/vs/workbench/api/electron-browser/mainThreadSCM.ts [hunk 1]
```diff
@@ -273,7 +275,10 @@ export class MainThreadSCM implements MainThreadSCMShape {
 
 		this._disposables.push(this.scmService.onDidChangeSelectedRepositories(repositories => {
 			const handles = repositories.map(repo => (<MainThreadSCMProvider>repo.provider).handle);
-			this._proxy.$onDidChangeVisibleSourceControls(handles);
+			if (!equals(handles, this._selectedRepositories)) {
+				this._selectedRepositories = handles;
+				this._proxy.$onDidChangeVisibleSourceControls(handles);
+			}
```
Comment (joaomoreno) [inline src/vs/workbench/api/electron-browser/mainThreadSCM.ts]: :clap:

File: src/vs/vscode.proposed.d.ts [hunk 2]
```diff
@@ -1045,4 +1045,20 @@ declare module 'vscode' {
...
+		 */
+		export const visibleSourceControls: SourceControl[];
+
+		/**
+		 * An [event](#Event) which fires when the visible [source controls](#SourceControl) have changed.
+		 */
+		export const onDidChangeVisibleSourceControls: Event<SourceControl[]>;
+	}
+
+	//#endregion
```
Comment (joaomoreno) [inline src/vs/vscode.proposed.d.ts]: We should actually expose readonly versions of `SourceControl` instances. We don't want other extensions to be able to modify SourceControl instances.

**Review by joaomoreno — CHANGES_REQUESTED** [review]
We should actually expose readonly versions of `SourceControl` instances. We don't want other extensions to be able to modify SourceControl instances.

File: src/vs/workbench/api/node/extHostSCM.ts [hunk 3]
```diff
@@ -577,7 +578,16 @@ export class ExtHostSCM implements ExtHostSCMShape {
 		this.logService.trace('ExtHostSCM#$onVisibleSouceControlsChange', sourceControlHandles);
 
 		const sourceControls = sourceControlHandles
-			.map(handle => this._sourceControls.get(handle))
+			.map(handle => {
+				const sourceControl = this._sourceControls.get(handle);
+				return sourceControl
+					? deepFreeze({
```
Comment (joaomoreno) [inline src/vs/workbench/api/node/extHostSCM.ts]: Don't think the `deepFreeze` is really necessary. :+1:

File: src/vs/vscode.proposed.d.ts [hunk 4]
```diff
@@ -1048,16 +1048,35 @@ declare module 'vscode' {
 
 	//#region scm, Rachel & Peng https://github.com/Microsoft/vscode/issues/53892
 
+	export interface ReadonlySourceControl {
```
Comment (joaomoreno) [inline src/vs/vscode.proposed.d.ts]: How about:

- `SourceControlReference`
- `ActiveSourceControl`
- `VisibleSourceControl`
- ... hmmm

@jrieken For more ideas.

**Review by joaomoreno — CHANGES_REQUESTED** [review]

File: src/vs/vscode.proposed.d.ts (hunk 4 above)
Comment (jrieken) [reply src/vs/vscode.proposed.d.ts]: Why not just hand out references to `SourceControl` straight away? Are we scared that another extension does bag things with it?

Comment (joaomoreno) [reply src/vs/vscode.proposed.d.ts]: Yeah, those instances are the gate way to create groups, change group resources, etc... and there would be no way of preventing extensions from doing that.

In order to get that kind of functionality, I'd prefer opening up an official Git extension API, which I already started doing.

**Comment by joaomoreno** [comment]
Closing in favour of https://github.com/Microsoft/vscode/issues/53892#issuecomment-406325134

**Pull request closed without merging**